pkg/services/ngalert/migration/service.go

```diff
@@ -49,9 +69,464 @@ func ProvideService(
+	return ms.verifyTry(ctx, orgID, func(ctx context.Context) (*definitions.OrgMigrationSummary, error) {
+		summary := definitions.OrgMigrationSummary{}
+		om := ms.newOrgMigration(orgID)
+		alerts, err := ms.migrationStore.GetDashboardAlerts(ctx, orgID, dashboardID)
+		if err != nil {
+			return nil, fmt.Errorf("get alerts: %w", err)
+		}
+
+		du := om.migrateDashboard(ctx, dashboardID, alerts)
+		s, err := ms.newSync(orgID).syncAndSaveState(ctx, []*migmodels.DashboardUpgrade{du}, nil, skipExisting)
+		if err != nil {
+			return nil, err
+		}
+
+		summary.Add(s)
+		return &summary, nil
+	})
+}
+
+// MigrateAllDashboardAlerts migrates all legacy alerts to unified alerting contact points.
+func (ms *migrationService) MigrateAllDashboardAlerts(ctx context.Context, orgID int64, skipExisting bool) (definitions.OrgMigrationSummary, error) {
+	return ms.verifyTry(ctx, orgID, func(ctx context.Context) (*definitions.OrgMigrationSummary, error) {
+		summary := definitions.OrgMigrationSummary{}
+		om := ms.newOrgMigration(orgID)
+		dashboardUpgrades, err := om.migrateOrgAlerts(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		s, err := ms.newSync(orgID).syncAndSaveState(ctx, dashboardUpgrades, nil, skipExisting)
+		if err != nil {
+			return nil, err
+		}
+
+		summary.Add(s)
+		return &summary, nil
+	})
+}
+
+// MigrateOrg executes the migration for a single org.
+func (ms *migrationService) MigrateOrg(ctx context.Context, orgID int64, skipExisting bool) (definitions.OrgMigrationSummary, error) {
+	return ms.try(ctx, func(ctx context.Context) (*definitions.OrgMigrationSummary, error) {
+		summary := definitions.OrgMigrationSummary{}
+		ms.log.Info("Starting legacy migration for org", "orgID", orgID, "skipExisting", skipExisting)
+		om := ms.newOrgMigration(orgID)
+		dashboardUpgrades, pairs, err := om.migrateOrg(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		s, err := ms.newSync(orgID).syncAndSaveState(ctx, dashboardUpgrades, pairs, skipExisting)
+		if err != nil {
+			return nil, err
+		}
+
+		err = ms.migrationStore.SetMigrated(ctx, orgID, true)
+		if err != nil {
+			return nil, fmt.Errorf("setting migration status: %w", err)
+		}
+
+		summary.Add(s)
+		return &summary, nil
+	})
+}
+
+// GetOrgMigrationState returns the current migration state for an org. This is a potentially expensive operation as it
+// requires re-hydrating the entire migration state from the database against all current alerting resources.
+func (ms *migrationService) GetOrgMigrationState(ctx context.Context, orgID int64) (*definitions.OrgMigrationState, error) {
+	if migrated, err := ms.migrationStore.IsMigrated(ctx, orgID); err != nil || !migrated {
+		return &definitions.OrgMigrationState{OrgID: orgID}, err
+	}
+	var resState *definitions.OrgMigrationState
+	err := ms.store.InTransaction(ctx, func(ctx context.Context) error {
+		var err error
```

Comment:
I think the body of this transaction could be extracted to a function\method and split apart. It makes linter nervous.